syntax = "proto3";
package Proto;

// import "google/protobuf/timestamp.proto";
import "Exchange.proto";
import "Strategy.proto";
import "Request.proto";
import "Reply.proto";

enum Side {
  Buy = 0;
  Sell = 1;
  BuyCover = 2;
  BuyCoverToday = 3;
  BuyCoverYesterday = 4;
  SellCover = 5;
  SellCoverToday = 6;
  SellCoverYesterday = 7;
}

enum TimeCondition {
  GTD = 0;
  IOC = 1;
}

enum OrderType {
  Limit = 0;
  Market = 1;
}

enum OrderStatus {
  Undefined = 0;
  Local = 1;
  Submitted = 2;
  New = 3;
  PartialFilled = 4;
  Filled = 5;
  PartialFilledCanceled = 6;
  Canceled = 7;
  Rejected = 8;
}

enum OrderAction {
  Submit = 0;
  Cancel = 1;
  Amend = 2;
}

message Order {
  uint64 id = 1;
  string instrument = 2;
  string counter_id = 3;
  string exchange_id = 4;
  string quote_id = 5;
  string strategy = 6;
  string note = 7;
  double spot = 8;
  double theo = 9;
  double delta = 10;
  double volatility = 11;
  double ss_rate = 12;
  double credit = 13;
  double price = 14;
  double avg_executed_price = 15;
  int32 volume = 16;
  int32 executed_volume = 17;
  Exchange exchange = 18;
  StrategyType strategy_type = 19;
  Side side = 20;
  TimeCondition time_condition = 21;
  OrderType type = 22;
  OrderStatus status = 23;
  uint64 time = 24;
  int32 latency = 25;
}

message OrderRequest {
  OrderAction action = 1;
  repeated Order orders = 2;
}

message OrderReq {
  RequestType type = 1;
  Exchange exchange = 2;
  string user = 3;
  repeated Order orders = 4;
}

message OrderRep {
  repeated Order orders = 1;
  Reply result = 2;
}
